x-pack/plugins/lens/public/heatmap_visualization/chart_component.tsx

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { FC, useEffect, useState } from 'react';
+import {
+  Chart,
+  ElementClickListener,
+  Heatmap,
+  HeatmapBrushEvent,
+  HeatmapElementEvent,
+  HeatmapSpec,
+  ScaleType,
+  Settings,
+} from '@elastic/charts';
+import { euiPaletteForTemperature } from '@elastic/eui';
+import { VisualizationContainer } from '../visualization_container';
+import { HeatmapRenderProps } from './types';
+import './index.scss';
+import { LensBrushEvent, LensFilterEvent } from '../types';
+import { desanitizeFilterContext } from '../utils';
+import { search } from '../../../../../src/plugins/data/public';
+import { EmptyPlaceholder } from '../shared_components';
+import { LensIconChartHeatmap } from '../assets/chart_heatmap';
+
+declare global {
+  interface Window {
+    /**
+     * Flag used to enable debugState on elastic charts
+     */
+    _echDebugStateFlag?: boolean;
+  }
+}
+
+export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+  data,
+  args,
+  timeZone,
+  formatFactory,
+  chartsThemeService,
+  onClickValue,
+  onSelectRange,
+}) => {
+  const chartTheme = chartsThemeService.useChartsTheme();
+  const isDarkTheme = chartsThemeService.useDarkMode();
+
+  const tableId = Object.keys(data.tables)[0];
+  const table = data.tables[tableId];
+
+  const xAxisColumnIndex = table.columns.findIndex((v) => v.id === args.xAccessor);
+  const yAxisColumnIndex = table.columns.findIndex((v) => v.id === args.yAccessor);
+
+  const xAxisColumn = table.columns[xAxisColumnIndex];
+  const yAxisColumn = table.columns[yAxisColumnIndex];
+  const valueColumn = table.columns.find((v) => v.id === args.valueAccessor);
+
+  if (!xAxisColumn || !valueColumn) {
+    // Chart is not ready
+    return null;
+  }
+
+  let chartData = table.rows.filter((v) => typeof v[args.valueAccessor!] === 'number');
+
+  if (!yAxisColumn) {
+    // required for tooltip
+    chartData = chartData.map((row) => {
+      return {
+        ...row,
+        unifiedY: '',
+      };
+    });
+  }
+
+  const xAxisMeta = xAxisColumn.meta;
+  const isTimeBasedSwimLane = xAxisMeta.type === 'date';
+
+  // Fallback to the ordinal scale type when a single row of data is provided.
+  // Related issue https://github.com/elastic/elastic-charts/issues/1184
+  const xScaleType =
+    isTimeBasedSwimLane && chartData.length > 1 ? ScaleType.Time : ScaleType.Ordinal;
+
+  const xValuesFormatter = formatFactory(xAxisMeta.params);
+  const valueFormatter = formatFactory(valueColumn.meta.params);
+
+  // Enable xDomain when https://github.com/elastic/elastic-charts/issues/1165 is resolved.
+  // @ts-ignore
```

Comment:
Nit: We usually don't keep these thins in code, can you remove it from here and move it into an issue?

Reply:
Removed in https://github.com/elastic/kibana/pull/97978/commits/9bb03dfeaf3df3d7a30c9449c8a7684ca50882c4